### PR TITLE
Speed up BundleFuse

### LINKED ISSUE #757: Make cl mount faster
Currently the REST calls dominate the processing times for every read request. Try to find a way to buffer.

### PR DESCRIPTION
**Pull request by fabeschan**
Prefetch byte ranges.

Fixes #757

### CODE REVIEW

**Comment by fabeschan** [comment]
I'll do a clean up once I get an approval for the overall approach

File: codalab/lib/bundle_fuse.py
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Add high-level description.

File: codalab/lib/bundle_fuse.py
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Probably don't want these print statements, or guard them by some verbose flag.

**Review by percyliang — APPROVED** [review]
Looks generally reasonable.  In general, more documentation, especially about the performance characteristics about how the code is exercised in practice, would be nice.

Question: if you do a cp of a 30 MB file, what is `length` usually?  If this is larger than 1 MB, then we could have one call that grabs multiple chunks.  It seems like the REST API is still a bottleneck given the difference between using FUSE and `cl down`.  Probably worth quantifying this.

**Comment by fabeschan** [comment]
Thanks @percyliang , I'll do a cleanup and add documentation. 

`length` is usually around ~13000 bytes

File: codalab/lib/bundle_fuse.py
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Say that partial chunks are only at the end of the file because they could grow...

Make the assumptions clear.

Comment (fabeschan) [reply codalab/lib/bundle_fuse.py]: done

File: codalab/lib/bundle_fuse.py
Comment (fabeschan) [reply codalab/lib/bundle_fuse.py]: done

**Comment by maxwang7** [comment]
LGTM! I didn't do any perf testing, but functionally it works for me on Ubuntu 14.04.

**Pull request merged**
